pypeit/images/rawimage.py

```diff
@@ -19,399 +19,701 @@
+        # NOTE: Prevent estimate_readnoise() from altering self.detector using
+        # deepcopy
+        self.ronoise = deepcopy(self.detector['ronoise'])
 
         # Attributes
         self.par = None
         self.ivar = None
+        self.var = None
         self.rn2img = None
         self.spat_flexure_shift = None
+        self._bpm = None
 
-        # All possible processing steps
-        #  Note these have to match the method names below
-        self.steps = dict(subtract_bias=False,
+        # All possible processing steps.  NOTE: These have to match the
+        # method/function names.  Their order here matches there execution order
+        # in self.process(), but that's not necessary.
+        self.steps = dict(subtract_pattern=False,
                           subtract_overscan=False,
-                          subtract_dark=False,
-                          subtract_pattern=False,
                           trim=False,
                           orient=False,
+                          subtract_bias=False,
                           apply_gain=False,
-                          flatten=False,
-                          )
+                          subtract_dark=False,
+                          spatial_flexure_shift=False,
+                          flatten=False)
 
     @property
     def bpm(self):
         """
-        Generate and return the bad pixel mask for this image
-        Warning:  BPM masks are for processed (e.g. trimmed, rotated) images only!
+        Generate and return the bad pixel mask for this image.
+
+        .. warning::
+
+            BPMs are for processed (e.g. trimmed, rotated) images only!
 
         Returns:
             `numpy.ndarray`_:  Bad pixel mask with a bad pixel = 1
 
         """
         if self._bpm is None:
-            self._bpm = self.spectrograph.bpm(shape=self.image.shape,
-                                    filename=self.filename,
-                                    det=self.det)
+            # TODO: Was this ever being called?
+#            self._bpm = self.spectrograph.bpm(shape=self.image.shape, filename=self.filename,
+#                                              det=self.det)
+            # TODO: Pass bias if there is one?
+            self._bpm = self.spectrograph.bpm(self.filename, self.det, shape=self.image.shape)
         return self._bpm
 
-    def apply_gain(self, force=False):
+    @property
+    def use_flat(self):
         """
-        Apply the Gain values to self.image
+        Return a flag setting if the flat data should be used in the image
+        processing.
+        """
+        if self.par is None:
+            return False
+        # TODO: At the moment, we can only perform any of the flat-field
+        # corrections if we are applying the pixel-flat correction.
+#        return self.par['use_pixelflat'] or self.par['use_specillum'] or self.par['use_illumflat']
+        return self.par['use_pixelflat']
```

Comment:
return self.par['use_pixelflat'] if self.par is not None else False

Reply:
Good call, but will there ever be a case where someone will be allowed to apply, say, use_illumflat without use_pixelflat?  If so, I'd rather leave it as is.  Otherwise, I'll make the change.